### ISSUE TITLE
compile scala and java project , Exception  java.lang.StackOverflowError

### ISSUE THREAD

**Issue by FireFreedomK** [davidB/scala-maven-plugin#641]
Compiling successfully with IDEA will fail with Maven

```
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  27.847 s
[INFO] Finished at: 2022-09-08T17:31:14+08:00
[INFO] ------------------------------------------------------------------------
---------------------------------------------------
constituent[0]: file:/C:/main/apache-maven-3.8.3/conf/logging/
constituent[1]: file:/C:/main/apache-maven-3.8.3/lib/commons-cli-1.4.jar
constituent[2]: file:/C:/main/apache-maven-3.8.3/lib/commons-io-2.6.jar
constituent[3]: file:/C:/main/apache-maven-3.8.3/lib/commons-lang3-3.8.1.jar
constituent[4]: file:/C:/main/apache-maven-3.8.3/lib/guava-25.1-android.jar
constituent[5]: file:/C:/main/apache-maven-3.8.3/lib/guice-4.2.2-no_aop.jar
constituent[6]: file:/C:/main/apache-maven-3.8.3/lib/jansi-2.3.4.jar
constituent[7]: file:/C:/main/apache-maven-3.8.3/lib/javax.annotation-api-1.2.jar
constituent[8]: file:/C:/main/apache-maven-3.8.3/lib/javax.inject-1.jar
constituent[9]: file:/C:/main/apache-maven-3.8.3/lib/jcl-over-slf4j-1.7.32.jar
constituent[10]: file:/C:/main/apache-maven-3.8.3/lib/maven-artifact-3.8.3.jar
constituent[11]: file:/C:/main/apache-maven-3.8.3/lib/maven-builder-support-3.8.3.jar
constituent[12]: file:/C:/main/apache-maven-3.8.3/lib/maven-compat-3.8.3.jar
constituent[13]: file:/C:/main/apache-maven-3.8.3/lib/maven-core-3.8.3.jar
constituent[14]: file:/C:/main/apache-maven-3.8.3/lib/maven-embedder-3.8.3.jar
constituent[15]: file:/C:/main/apache-maven-3.8.3/lib/maven-model-3.8.3.jar
constituent[16]: file:/C:/main/apache-maven-3.8.3/lib/maven-model-builder-3.8.3.jar
constituent[17]: file:/C:/main/apache-maven-3.8.3/lib/maven-plugin-api-3.8.3.jar
constituent[18]: file:/C:/main/apache-maven-3.8.3/lib/maven-repository-metadata-3.8.3.jar
constituent[19]: file:/C:/main/apache-maven-3.8.3/lib/maven-resolver-api-1.6.3.jar
constituent[20]: file:/C:/main/apache-maven-3.8.3/lib/maven-resolver-connector-basic-1.6.3.jar
constituent[21]: file:/C:/main/apache-maven-3.8.3/lib/maven-resolver-impl-1.6.3.jar
constituent[22]: file:/C:/main/apache-maven-3.8.3/lib/maven-resolver-provider-3.8.3.jar
constituent[23]: file:/C:/main/apache-maven-3.8.3/lib/maven-resolver-spi-1.6.3.jar
constituent[24]: file:/C:/main/apache-maven-3.8.3/lib/maven-resolver-transport-wagon-1.6.3.jar
constituent[25]: file:/C:/main/apache-maven-3.8.3/lib/maven-resolver-util-1.6.3.jar
constituent[26]: file:/C:/main/apache-maven-3.8.3/lib/maven-settings-3.8.3.jar
constituent[27]: file:/C:/main/apache-maven-3.8.3/lib/maven-settings-builder-3.8.3.jar
constituent[28]: file:/C:/main/apache-maven-3.8.3/lib/maven-shared-utils-3.3.4.jar
constituent[29]: file:/C:/main/apache-maven-3.8.3/lib/maven-slf4j-provider-3.8.3.jar
constituent[30]: file:/C:/main/apache-maven-3.8.3/lib/org.eclipse.sisu.inject-0.3.5.jar
constituent[31]: file:/C:/main/apache-maven-3.8.3/lib/org.eclipse.sisu.plexus-0.3.5.jar
constituent[32]: file:/C:/main/apache-maven-3.8.3/lib/plexus-cipher-2.0.jar
constituent[33]: file:/C:/main/apache-maven-3.8.3/lib/plexus-component-annotations-2.1.0.jar
constituent[34]: file:/C:/main/apache-maven-3.8.3/lib/plexus-interpolation-1.26.jar
constituent[35]: file:/C:/main/apache-maven-3.8.3/lib/plexus-sec-dispatcher-2.0.jar
constituent[36]: file:/C:/main/apache-maven-3.8.3/lib/plexus-utils-3.3.0.jar
constituent[37]: file:/C:/main/apache-maven-3.8.3/lib/slf4j-api-1.7.32.jar
constituent[38]: file:/C:/main/apache-maven-3.8.3/lib/wagon-file-3.4.3.jar
constituent[39]: file:/C:/main/apache-maven-3.8.3/lib/wagon-http-3.4.3-shaded.jar
constituent[40]: file:/C:/main/apache-maven-3.8.3/lib/wagon-provider-api-3.4.3.jar
---------------------------------------------------
Exception in thread "main" java.lang.StackOverflowError
        at scala.tools.nsc.transform.Erasure$Eraser.adaptMember(Erasure.scala:690)
        at scala.tools.nsc.transform.Erasure$Eraser.typed1(Erasure.scala:809)
        at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5781)
        at scala.tools.nsc.transform.Erasure$Eraser.adaptMember(Erasure.scala:734)
        at scala.tools.nsc.transform.Erasure$Eraser.typed1(Erasure.scala:809)
        at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5781)
        at scala.tools.nsc.typechecker.Typers$Typer.$anonfun$typed1$39(Typers.scala:4891)
        at scala.tools.nsc.typechecker.Typers$Typer.silent(Typers.scala:713)
        at scala.tools.nsc.typechecker.Typers$Typer.normalTypedApply$1(Typers.scala:4893)
        at scala.tools.nsc.typechecker.Typers$Typer.typedApply$1(Typers.scala:4921)
        at scala.tools.nsc.typechecker.Typers$Typer.typed1(Typers.scala:5736)
        at scala.tools.nsc.transform.Erasure$Eraser.typed1(Erasure.scala:809)
        at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5781)
        at scala.tools.nsc.transform.Erasure$Eraser.adaptMember(Erasure.scala:734)
        at scala.tools.nsc.transform.Erasure$Eraser.typed1(Erasure.scala:809)
        at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5781)
        at scala.tools.nsc.typechecker.Typers$Typer.$anonfun$typed1$39(Typers.scala:4891)
        at scala.tools.nsc.typechecker.Typers$Typer.silent(Typers.scala:713)
        at scala.tools.nsc.typechecker.Typers$Typer.normalTypedApply$1(Typers.scala:4893)
        at scala.tools.nsc.typechecker.Typers$Typer.typedApply$1(Typers.scala:4921)
        at scala.tools.nsc.typechecker.Typers$Typer.typed1(Typers.scala:5736)
        at scala.tools.nsc.transform.Erasure$Eraser.typed1(Erasure.scala:809)
        at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5781)
        at scala.tools.nsc.transform.Erasure$Eraser.adaptMember(Erasure.scala:734)
        at scala.tools.nsc.transform.Erasure$Eraser.typed1(Erasure.scala:809)
        at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5781)
        at scala.tools.nsc.typechecker.Typers$Typer.$anonfun$typed1$39(Typers.scala:4891)
        at scala.tools.nsc.typechecker.Typers$Typer.silent(Typers.scala:713)
        at scala.tools.nsc.typechecker.Typers$Typer.normalTypedApply$1(Typers.scala:4893)
        at scala.tools.nsc.typechecker.Typers$Typer.typedApply$1(Typers.scala:4921)
        at scala.tools.nsc.typechecker.Typers$Typer.typed1(Typers.scala:5736)
        at scala.tools.nsc.transform.Erasure$Eraser.typed1(Erasure.scala:809)
        at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5781)
        at scala.tools.nsc.transform.Erasure$Eraser.adaptMember(Erasure.scala:734)
        at scala.tools.nsc.transform.Erasure$Eraser.typed1(Erasure.scala:809)
        at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5781)
        at scala.tools.nsc.typechecker.Typers$Typer.$anonfun$typed1$39(Typers.scala:4891)
        at scala.tools.nsc.typechecker.Typers$Typer.silent(Typers.scala:713)
        at scala.tools.nsc.typechecker.Typers$Typer.normalTypedApply$1(Typers.scala:4893)
        at scala.tools.nsc.typechecker.Typers$Typer.typedApply$1(Typers.scala:4921)
        at scala.tools.nsc.typechecker.Typers$Typer.typed1(Typers.scala:5736)
        at scala.tools.nsc.transform.Erasure$Eraser.typed1(Erasure.scala:809)
        at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5781)
        at scala.tools.nsc.transform.Erasure$Eraser.adaptMember(Erasure.scala:734)
        at scala.tools.nsc.transform.Erasure$Eraser.typed1(Erasure.scala:809)
        at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5781)
        at scala.tools.nsc.typechecker.Typers$Typer.$anonfun$typed1$39(Typers.scala:4891)
        at scala.tools.nsc.typechecker.Typers$Typer.silent(Typers.scala:713)
        at scala.tools.nsc.typechecker.Typers$Typer.normalTypedApply$1(Typers.scala:4893)
        at scala.tools.nsc.typechecker.Typers$Typer.typedApply$1(Typers.scala:4921)
        at scala.tools.nsc.typechecker.Typers$Typer.typed1(Typers.scala:5736)
        at scala.tools.nsc.transform.Erasure$Eraser.typed1(Erasure.scala:809)
        at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5781)
        at scala.tools.nsc.transform.Erasure$Eraser.adaptMember(Erasure.scala:734)
        at scala.tools.nsc.transform.Erasure$Eraser.typed1(Erasure.scala:809)
        at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5781)
        at scala.tools.nsc.typechecker.Typers$Typer.$anonfun$typed1$39(Typers.scala:4891)
        at scala.tools.nsc.typechecker.Typers$Typer.silent(Typers.scala:713)
        at scala.tools.nsc.typechecker.Typers$Typer.normalTypedApply$1(Typers.scala:4893)
        at scala.tools.nsc.typechecker.Typers$Typer.typedApply$1(Typers.scala:4921)
        at scala.tools.nsc.typechecker.Typers$Typer.typed1(Typers.scala:5736)
        at scala.tools.nsc.transform.Erasure$Eraser.typed1(Erasure.scala:809)
        at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5781)
        at scala.tools.nsc.transform.Erasure$Eraser.adaptMember(Erasure.scala:734)
        at scala.tools.nsc.transform.Erasure$Eraser.typed1(Erasure.scala:809)
        at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5781)
        at scala.tools.nsc.typechecker.Typers$Typer.$anonfun$typed1$39(Typers.scala:4891)
        at scala.tools.nsc.typechecker.Typers$Typer.silent(Typers.scala:713)
        at scala.tools.nsc.typechecker.Typers$Typer.normalTypedApply$1(Typers.scala:4893)
        at scala.tools.nsc.typechecker.Typers$Typer.typedApply$1(Typers.scala:4921)
        at scala.tools.nsc.typechecker.Typers$Typer.typed1(Typers.scala:5736)
        at scala.tools.nsc.transform.Erasure$Eraser.typed1(Erasure.scala:809)
        at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5781)
        at scala.tools.nsc.transform.Erasure$Eraser.adaptMember(Erasure.scala:734)
        at scala.tools.nsc.transform.Erasure$Eraser.typed1(Erasure.scala:809)
        at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5781)
        at scala.tools.nsc.typechecker.Typers$Typer.$anonfun$typed1$39(Typers.scala:4891)
        at scala.tools.nsc.typechecker.Typers$Typer.silent(Typers.scala:713)
        at scala.tools.nsc.typechecker.Typers$Typer.normalTypedApply$1(Typers.scala:4893)
        at scala.tools.nsc.typechecker.Typers$Typer.typedApply$1(Typers.scala:4921)
        at scala.tools.nsc.typechecker.Typers$Typer.typed1(Typers.scala:5736)
        at scala.tools.nsc.transform.Erasure$Eraser.typed1(Erasure.scala:809)
        at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5781)
        at scala.tools.nsc.transform.Erasure$Eraser.adaptMember(Erasure.scala:734)
        at scala.tools.nsc.transform.Erasure$Eraser.typed1(Erasure.scala:809)
        at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5781)
        at scala.tools.nsc.typechecker.Typers$Typer.$anonfun$typed1$39(Typers.scala:4891)
        at scala.tools.nsc.typechecker.Typers$Typer.silent(Typers.scala:713)
        at scala.tools.nsc.typechecker.Typers$Typer.normalTypedApply$1(Typers.scala:4893)
        at scala.tools.nsc.typechecker.Typers$Typer.typedApply$1(Typers.scala:4921)
        at scala.tools.nsc.typechecker.Typers$Typer.typed1(Typers.scala:5736)
        at scala.tools.nsc.transform.Erasure$Eraser.typed1(Erasure.scala:809)
        at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5781)
        at scala.tools.nsc.transform.Erasure$Eraser.adaptMember(Erasure.scala:734)
        at scala.tools.nsc.transform.Erasure$Eraser.typed1(Erasure.scala:809)
        at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5781)
        at scala.tools.nsc.typechecker.Typers$Typer.$anonfun$typed1$39(Typers.scala:4891)
        at scala.tools.nsc.typechecker.Typers$Typer.silent(Typers.scala:713)
        at scala.tools.nsc.typechecker.Typers$Typer.normalTypedApply$1(Typers.scala:4893)
        at scala.tools.nsc.typechecker.Typers$Typer.typedApply$1(Typers.scala:4921)
        at scala.tools.nsc.typechecker.Typers$Typer.typed1(Typers.scala:5736)
        at scala.tools.nsc.transform.Erasure$Eraser.typed1(Erasure.scala:809)
        at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5781)
        at scala.tools.nsc.transform.Erasure$Eraser.adaptMember(Erasure.scala:734)
        at scala.tools.nsc.transform.Erasure$Eraser.typed1(Erasure.scala:809)
        at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5781)
        at scala.tools.nsc.typechecker.Typers$Typer.$anonfun$typed1$39(Typers.scala:4891)
        at scala.tools.nsc.typechecker.Typers$Typer.silent(Typers.scala:713)
        at scala.tools.nsc.typechecker.Typers$Typer.normalTypedApply$1(Typers.scala:4893)
        at scala.tools.nsc.typechecker.Typers$Typer.typedApply$1(Typers.scala:4921)
        at scala.tools.nsc.typechecker.Typers$Typer.typed1(Typers.scala:5736)
        at scala.tools.nsc.transform.Erasure$Eraser.typed1(Erasure.scala:809)
        at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5781)
        at scala.tools.nsc.transform.Erasure$Eraser.adaptMember(Erasure.scala:734)
        at scala.tools.nsc.transform.Erasure$Eraser.typed1(Erasure.scala:809)
        at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5781)
        at scala.tools.nsc.typechecker.Typers$Typer.$anonfun$typed1$39(Typers.scala:4891)
        at scala.tools.nsc.typechecker.Typers$Typer.silent(Typers.scala:713)
        at scala.tools.nsc.typechecker.Typers$Typer.normalTypedApply$1(Typers.scala:4893)
        at scala.tools.nsc.typechecker.Typers$Typer.typedApply$1(Typers.scala:4921)
        at scala.tools.nsc.typechecker.Typers$Typer.typed1(Typers.scala:5736)
        at scala.tools.nsc.transform.Erasure$Eraser.typed1(Erasure.scala:809)
        at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5781)
       ...
       ...
        at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5781)
        at scala.tools.nsc.transform.Erasure$Eraser.adaptMember(Erasure.scala:734)
```

My pom.xml file content:
```
<build>
        <plugins>
            <plugin>
                <groupId>net.alchim31.maven</groupId>
                <artifactId>scala-maven-plugin</artifactId>
                <version>4.7.1</version>
                <executions>
                    <execution>
                        <goals>
                            <goal>compile</goal>
                            <goal>testCompile</goal>
                        </goals>
                        <configuration>
                            <recompileMode>incremental</recompileMode>
                            <includes>
                                <include>**/*.scala</include>
                            </includes>
                        </configuration>
                    </execution>
                </executions>
            </plugin>

            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-compiler-plugin</artifactId>
                <version>3.0</version>
                <configuration>
                    <source>8</source>
                    <target>8</target>
                    <encoding>UTF-8</encoding>
                </configuration>
                <executions>
                    <execution>
                        <phase>compile</phase>
                        <goals>
                            <goal>compile</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>

            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-shade-plugin</artifactId>
                <version>3.2.4</version>
                <executions>
                    <!-- Run shade goal on package phase -->
                    <execution>
                        <phase>package</phase>
                        <goals>
                            <goal>shade</goal>
                        </goals>
                        <configuration>
                            <createDependencyReducedPom>false</createDependencyReducedPom>
                            <filters>
                                <filter>
                                    <!-- Do not copy the signatures in the META-INF folder.
                                    Otherwise, this might cause SecurityExceptions when using the JAR. -->
                                    <artifact>*:*</artifact>
                                    <excludes>
                                        <exclude>module-info.class</exclude>
                                        <exclude>META-INF/*.SF</exclude>
                                        <exclude>META-INF/*.DSA</exclude>
                                        <exclude>META-INF/*.RSA</exclude>
                                    </excludes>
                                </filter>
                            </filters>

                            <transformers>
                                <transformer implementation="org.apache.maven.plugins.shade.resource.ManifestResourceTransformer">
                                    <mainClass>${mainClass}</mainClass>
                                </transformer>
                            </transformers>

                            <createDependencyReducedPom>false</createDependencyReducedPom>
                        </configuration>
                    </execution>
                </executions>
            </plugin>

        </plugins>
    </build>
```
My idea maven vm options:
idea --> Settings --> Build,Execution,Deployment --> Build Tools --> Maven --> Runner
```
-Xmx4096m -XX:MetaspaceSize=1024m -XX:MaxMetaspaceSize=1024m -Xss32m -Dmaven.test.skip=true -Dmaven.compile.fork=true
```

**Comment by slandelle** [comment]
> My idea maven vm options: ...

Those are the setting for maven itself, not for the forked compiler process.

See https://davidb.github.io/scala-maven-plugin/example_compile.html "Increasing Memory".

**Comment by FireFreedomK** [comment]
@slandelle I have tried this method, but it still prompts the error.

**Comment by slandelle** [comment]
Please provide a proper reproducer: https://stackoverflow.com/help/minimal-reproducible-example

**Comment by slandelle** [comment]
This is probably a duplicate of #507

**Comment by slandelle** [comment]
Fixed in 4.8.0
